### ISSUE TITLE
ASSERTION FAILED: blocked_thread_list.contains(current_thread)

### ISSUE THREAD

**Issue by diversys** [SerenityOS/serenity#9401]
This is x86_64 on macOS Qemu. Opening Assistant crashes the kernel:
```
Assistant(32:32): Started thread "", tid = 33
[Assistant(32:33)]: ASSERTION FAILED: blocked_thread_list.contains(current_thread)
[Assistant(32:33)]: ./Kernel/Locking/Mutex.cpp:209 in void Kernel::Mutex::block(Kernel::Thread&, Kernel::Mutex::Mode, Kernel::ScopedSpinLock<Kernel::SpinLock<unsigned char> >&, u32)
[Assistant(32:33)]: KERNEL PANIC! :^(
[Assistant(32:33)]: Aborted
[Assistant(32:33)]: at ./Kernel/Arch/x86/common/CPU.cpp:29 in void abort()

```

**Comment by kemzeb** [comment]
Encountered the same problem as well, running it on Ubuntu ARM64:
```
Assistant(35:35): Started thread "", tid = 36
[Assistant(35:36)]: ASSERTION FAILED: blocked_thread_list.contains(current_thread)
[Assistant(35:36)]: ../../Kernel/Locking/Mutex.cpp:209 in void Kernel::Mutex::block(Kernel::Thread&, Kernel::Mutex::Mode, Kernel::SpinlockLocker<Kernel::Spinlock<unsigned char> >&, u32)
[Assistant(35:36)]: KERNEL PANIC! :^(
[Assistant(35:36)]: Aborted
[Assistant(35:36)]: at ../../Kernel/Arch/x86/common/CPU.cpp:28 in void abort()
[Assistant(35:36)]: Kernel + 0x00467ead  Kernel::__panic(char const*, unsigned int, char const*) +0xb8
[Assistant(35:36)]: Kernel + 0x00618fd8  illegal_instruction_asm_entry +0x0
[Assistant(35:36)]: Kernel + 0x00618df9  abort.localalias +0x0
[Assistant(35:36)]: Kernel + 0x003cb103  Kernel::Mutex::block(Kernel::Thread&, Kernel::LockMode, Kernel::SpinlockLocker<Kernel::Spinlock<unsigned char> >&, unsigned int) [clone .localalias] +0x225
[Assistant(35:36)]: Kernel + 0x003cd436  Kernel::Mutex::lock(Kernel::LockMode, Kernel::LockLocation const&) +0x77e
[Assistant(35:36)]: Kernel + 0x0027c03d  Kernel::Ext2FS::get_inode(Kernel::InodeIdentifier) const [clone .localalias] +0xbb
[Assistant(35:36)]: Kernel + 0x00285b52  Kernel::Ext2FSInode::lookup(AK::StringView) [clone .localalias] +0xb88
[Assistant(35:36)]: Kernel + 0x0033c9aa  Kernel::VirtualFileSystem::resolve_path_without_veil(AK::StringView, Kernel::Custody&, AK::RefPtr<Kernel::Custody, AK::RefPtrTraits<Kernel::Custody> >*, int, int) [clone .localalias] +0xd08
[Assistant(35:36)]: Kernel + 0x0033f0cb  Kernel::VirtualFileSystem::resolve_path(AK::StringView, Kernel::Custody&, AK::RefPtr<Kernel::Custody, AK::RefPtrTraits<Kernel::Custody> >*, int, int) [clone .localalias] +0x4b
[Assistant(35:36)]: Kernel + 0x00344534  Kernel::VirtualFileSystem::lookup_metadata(AK::StringView, Kernel::Custody&, int) +0x4e
[Assistant(35:36)]: Kernel + 0x0057ef16  Kernel::Process::sys$stat(AK::Userspace<Kernel::Syscall::SC_stat_params const*>) +0x13c4
[Assistant(35:36)]: Kernel + 0x004bdc1b  syscall_handler +0x1592
[Assistant(35:36)]: Kernel + 0x004bc04d  syscall_asm_entry +0x31
qemu-system-i386: terminating on signal 2
```

**Comment by timschumi** [comment]
I can also reproduce this on i686 quite often by running `env GIT_SSL_NO_VERIFY=true git clone https://github.com/SerenityOS/serenity`:

```
[git(39:50)]: ASSERTION FAILED: blocked_thread_list.contains(current_thread)
[git(39:50)]: ./Kernel/Locking/Mutex.cpp:215 in void Kernel::Mutex::block(Kernel::Thread&, Kernel::Mutex::Mode, Kernel::SpinlockLocker<Kernel::Spinlock>&, u32)
[git(39:50)]: KERNEL PANIC! :^(
[git(39:50)]: Aborted
[git(39:50)]: at ./Kernel/Arch/x86/common/CPU.cpp:33 in void abort()
[git(39:50)]: Kernel + 0x004c2283  Kernel::__panic(char const*, unsigned int, char const*) +0x11c
[git(39:50)]: Kernel + 0x00687b14  illegal_instruction_asm_entry +0x0
[git(39:50)]: Kernel + 0x006878cb  abort.localalias +0x0
[git(39:50)]: Kernel + 0x0041fabd  Kernel::Mutex::block(Kernel::Thread&, Kernel::LockMode, Kernel::SpinlockLocker<Kernel::Spinlock>&, unsigned int) [clone .localalias] +0x225
[git(39:50)]: Kernel + 0x00421e1d  Kernel::Mutex::lock(Kernel::LockMode, Kernel::LockLocation const&) +0x785
[git(39:50)]: Kernel + 0x002a3ac3  Kernel::Ext2FSInode::read_bytes(long long, unsigned long, Kernel::UserOrKernelBuffer&, Kernel::OpenFileDescription*) const [clone .localalias] +0x117
[git(39:50)]: Kernel + 0x002e8860  Kernel::InodeFile::read(Kernel::OpenFileDescription&, unsigned long long, Kernel::UserOrKernelBuffer&, unsigned long) [clone .localalias] +0x2f8
[git(39:50)]: Kernel + 0x00313b6a  Kernel::OpenFileDescription::read(Kernel::UserOrKernelBuffer&, unsigned long) +0x3f8
[git(39:50)]: Kernel + 0x005b0386  Kernel::Process::sys$read(int, AK::Userspace<unsigned char*>, unsigned long) +0xf58
[git(39:50)]: Kernel + 0x0051ec77  syscall_handler +0x16e6
[git(39:50)]: Kernel + 0x0051cf4f  syscall_asm_entry +0x31
```

**Comment by GMTA** [comment]
On x86_64, running ScummVM + Grim Fandango triggers this bug fairly often as well. I tracked it down to these lines:

https://github.com/SerenityOS/serenity/blob/98e65f71da8f525c0b3b29d0c2264090007e0d5c/Kernel/Locking/Mutex.cpp#L209-L216

`blocked_thread_list` is either `Mutex::m_blocked_threads_list_exclusive` or `::m_blocked_threads_list_shared`, but I've only see it fail with `m_blocked_threads_list_exclusive`. At line 215, `current_thread` is no longer in `m_blocked_threads_list_exclusive`. The intrusive's list `nnode.m_storage` is always set to `0x0000000000000000`, meaning the node was removed from the list.

For these thread lists, the only times when a node gets removed, is at lines 209 (just before appending) and 216. That would implicate that either the same thread is executing `Mutex::block` twice, or that some other thread is claiming to be the wrong `current_thread`...

**Comment by csm-yujinkim** [comment]
Same issue when trying to change background and screen resolution at the same time:
<img width="1136" alt="Screen Shot 2021-12-05 at 6 00 19 PM" src="https://user-images.githubusercontent.com/60201675/144771735-d99e2091-83ef-47fd-9ad4-f6125529a3dd.png">

**Comment by GMTA** [comment]
Another trace from when data is being written:
```
[scummvm(35:37)]: ASSERTION FAILED: blocked_thread_list.contains(current_thread)
[scummvm(35:37)]: ../../Kernel/Locking/Mutex.cpp:215 in void Kernel::Mutex::block(Kernel::Thread&, Kernel::Mutex::Mode, Kernel::SpinlockLocker<Kernel::Spinlock>&, u32)
[scummvm(35:37)]: KERNEL PANIC! :^(
[scummvm(35:37)]: Aborted
[scummvm(35:37)]: at ../../Kernel/Arch/x86/common/CPU.cpp:35 in void abort()
[scummvm(35:37)]: Kernel + 0x0000000000a02910  Kernel::__panic(char const*, unsigned int, char const*) +0x1b6
[scummvm(35:37)]: Kernel + 0x0000000000c97835  abort.localalias +0x374
[scummvm(35:37)]: Kernel + 0x0000000000c974c1  abort.localalias +0x0
[scummvm(35:37)]: Kernel + 0x0000000000909ad4  Kernel::Mutex::block(Kernel::Thread&, Kernel::LockMode, Kernel::SpinlockLocker<Kernel::Spinlock>&, unsigned int) [clone .localalias] +0x2ec
[scummvm(35:37)]: Kernel + 0x000000000090ce13  Kernel::Mutex::lock(Kernel::LockMode, Kernel::LockLocation const&) +0xb0d
[scummvm(35:37)]: Kernel + 0x000000000066b1d0  Kernel::BlockBasedFileSystem::write_block(AK::DistinctNumeric<unsigned long, Kernel::__BlockIndex_tag, false, true, true, false, false, false>, Kernel::UserOrKernelBuffer const&, unsigned long, unsigned long, bool) [clone .localalias] +0x3f4
[scummvm(35:37)]: Kernel + 0x00000000006d2a18  Kernel::Ext2FSInode::write_bytes(long, unsigned long, Kernel::UserOrKernelBuffer const&, Kernel::OpenFileDescription*) [clone .localalias] +0x16f4
[scummvm(35:37)]: Kernel + 0x00000000006d0a0c  Kernel::Ext2FSInode::resize(unsigned long) [clone .localalias] +0x1c8a
[scummvm(35:37)]: Kernel + 0x00000000006d1f2d  Kernel::Ext2FSInode::write_bytes(long, unsigned long, Kernel::UserOrKernelBuffer const&, Kernel::OpenFileDescription*) [clone .localalias] +0xc09
[scummvm(35:37)]: Kernel + 0x000000000072d927  Kernel::InodeFile::write(Kernel::OpenFileDescription&, unsigned long, Kernel::UserOrKernelBuffer const&, unsigned long) [clone .localalias] +0x2d7
[scummvm(35:37)]: Kernel + 0x0000000000773ff6  Kernel::OpenFileDescription::write(Kernel::UserOrKernelBuffer const&, unsigned long) +0x432
[scummvm(35:37)]: Kernel + 0x0000000000bd3e54  Kernel::Process::do_write(Kernel::OpenFileDescription&, Kernel::UserOrKernelBuffer const&, unsigned long) [clone .localalias] +0x806
[scummvm(35:37)]: Kernel + 0x0000000000bd6303  Kernel::Process::sys$write(int, AK::Userspace<unsigned char const*>, unsigned long) +0x7d5
[scummvm(35:37)]: Kernel + 0x0000000000a96550  syscall_handler +0x1e08
[scummvm(35:37)]: Kernel + 0x0000000000a9380a  syscall_asm_entry +0x30
```

**Comment by GMTA** [comment]
This might have been fixed by commit b28beb691eea0cf4dbdebc14fa7e5b360c846383 - testing needed!

**Comment by GMTA** [comment]
Should be fixed by 7826729ab2ddbc087d5995522476291f75987132.
